medmodels/data_synthesis/mtgan/model/real_gru/real_gru.py

```diff
@@ -0,0 +1,123 @@
+            hyperparameters: The hyperparameters for the training.
+            device: The device to use for the model.
+        """
+        super().__init__()
+        self.total_number_of_concepts = total_number_of_concepts
+        self.epochs = hyperparamaters["real_gru_training_epochs"]
+        self.device = device
+        self.to(self.device)
+
+        self.gru = nn.GRU(
+            input_size=total_number_of_concepts,
+            hidden_size=hyperparamaters["generator_hidden_dimension"],
+            batch_first=True,
+            dtype=torch.float32,
+        )
+        self.linear_layer = nn.Sequential(
+            nn.Linear(
+                hyperparamaters["generator_hidden_dimension"], total_number_of_concepts
+            ),
+            nn.Sigmoid(),
+        )
+        self.optimizer = torch.optim.adam.Adam(
+            self.parameters(), lr=hyperparamaters["real_gru_learning_rate"]
+        )
+        self.loss_function = PredictionLoss(device=self.device)
+
+    def forward(self, data: torch.Tensor) -> torch.Tensor:
+        """Forward pass of RealGRU. Linear layer for first hidden state, then GRU.
+
+        Args:
+            x (torch.Tensor): input data, of shape (batch size, maximum number of
+                windows, total number of concepts)
+        Returns:
+            torch.Tensor: output of the RealGRU, of shape (batch size, maximum number
+                of windows , total number of concepts)
+        """
+        outputs, _ = self.gru(data)
+        return self.linear_layer(outputs)
+
+    def calculate_hidden_states(
+        self, data: torch.Tensor, number_of_windows_per_patient: torch.Tensor
+    ) -> torch.Tensor:
+        """Calculate hidden states with respect to the data and mask them according to the number of windows per patient.
+
+        Args:
+            data (torch.Tensor): input data, of shape (batch size, maximum number of
+                windows, total number of concepts)
+            number_of_windows_per_patient (torch.Tensor): number of windows per
+                patient
+
+        Returns:
+            torch.Tensor: hidden state of the RealGRU with respect to the data, masked
+        """
+        with torch.no_grad():
+            maximum_number_of_windows = data.shape[1]
+            mask = find_sequence_mask(
+                number_of_windows_per_patient, maximum_number_of_windows
+            ).unsqueeze(dim=-1)
+            outputs, _ = self.gru(data)
+            return outputs * mask
+
+    def train_real_gru(self, train_loader: MTGANDataLoader) -> float:
+        """Training the RealGRU model.
+
+        Args:
+            train_loader (MTGANDataLoader): The data loader for the training data.
+
+        Returns:
+            float: The loss of the model.
+        """
+        torch.manual_seed(0)
+        loss = torch.Tensor([np.inf])
+
+        for _ in range(1, self.epochs + 1):
+            for _, data in enumerate(train_loader, start=1):
```

Comment:
why enumerate starting with 1 if you throw the index away anyways?

Reply:
also the line above is the same as `for _ in range(0, self.epochs)`, no? (if the output is ignored via `_`)